### PR TITLE
Copy unique token store model into hedera mirror web3

### LINKED ISSUE #5757: HIP-584: Copy `UniqueToken` store model into `hedera-mirror-web3`
### Problem

We need to copy the `UniqueToken` store model into mirror-node, so that we use it as a value in the `Stacked State Frame` design. This would also mean the model should be immutable. It will both hold data for the `UniqueToken` and also perform some business logic internally.

### Solution

Copy `UniqueToken` store model into hedera-mirror-web3.

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by ivan-02**
**Description**:

We need to copy the `UniqueToken` store model into mirror-node, so that we use it as a value in the Stacked State Frame design. This would also mean the model should be immutable. It will both hold data for the `UniqueToken` and also perform some business logic internally.

**Related issue(s)**:

Fixes #5757

**Notes for reviewer**:

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/models/UniqueToken.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+ * Encapsulates the state and operations of a Hedera Unique token.
+ *
+ * <p>Operations are validated, and throw a {@link InvalidTransactionException} with response code
+ * capturing the failure when one occurs.
+ */
+@Getter
+@AllArgsConstructor
+@Builder(toBuilder = true)
+public class UniqueToken {
+    private Id tokenId;
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/services/store/models/UniqueToken.java:36]: You can add also an `Address` field that is initialized by `id.asEvmAddress()` in the constructor.

File: hedera-mirror-web3/src/main/java/com/hedera/services/utils/EntityNum.java [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+//        return STATIC_PROPERTIES.scopedIdLiteralWith(numFromCode(value));
+//    }
+//
+//    public Address toEvmAddress() {
+//        return Address.wrap(Bytes.wrap(toRawEvmAddress()));
+//    }
+//
+//    public byte[] toRawEvmAddress() {
+//        return asEvmAddress(numFromCode(value));
+//    }
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/services/utils/EntityNum.java:157]: If not used - delete all comments

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/models/UniqueToken.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import com.hedera.services.state.submerkle.RichInstant;
+
+/**
+ * Encapsulates the state and operations of a Hedera Unique token.
+ *
+ * <p>Operations are validated, and throw a {@link InvalidTransactionException} with response code
+ * capturing the failure when one occurs.
+ */
+@Getter
+@AllArgsConstructor
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/services/store/models/UniqueToken.java:33]: You can delete `@Getter` and `@AllArgsConstructor` and use `@Value` instead.

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/models/UniqueToken.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+import com.hedera.node.app.service.evm.exceptions.InvalidTransactionException;
+import com.hedera.services.state.submerkle.RichInstant;
+
+/**
+ * Encapsulates the state and operations of a Hedera Unique token.
+ *
+ * <p>Operations are validated, and throw a {@link InvalidTransactionException} with response code
+ * capturing the failure when one occurs.
+ */
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/services/store/models/UniqueToken.java:31]: We can enhance the javadoc with information that we explicitly make this class immutable and explain why.

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/models/UniqueToken.java (hunk 1 above)
Comment (ivan-02) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/models/UniqueToken.java:36]: Done

File: hedera-mirror-web3/src/main/java/com/hedera/services/utils/EntityNum.java (hunk 2 above)
Comment (ivan-02) [reply hedera-mirror-web3/src/main/java/com/hedera/services/utils/EntityNum.java:157]: Done

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/models/UniqueToken.java (hunk 3 above)
Comment (ivan-02) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/models/UniqueToken.java:35]: The `@Builder` requires all args constructor if we are using custom constructor - in this case the one with `Address` field initialization.

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/models/UniqueToken.java (hunk 4 above)
Comment (ivan-02) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/models/UniqueToken.java:33]: Done

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/models/UniqueToken.java (hunk 3 above)
Comment (IvanKavaldzhiev) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/models/UniqueToken.java:35]: I think we can delete the custom constructor and use the one provided by the @Value annotation.

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/models/UniqueToken.java (hunk 4 above)
Comment (IvanKavaldzhiev) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/models/UniqueToken.java:33]: We can adapt the last sentence like this:

This model is used as a value in a special state, used for speculative write operations.

File: hedera-mirror-web3/src/main/java/com/hedera/services/utils/EntityNum.java [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+import edu.umd.cs.findbugs.annotations.NonNull;
+import org.hyperledger.besu.datatypes.Address;
+
+import com.hedera.services.store.models.Id;
+
+/**
+ * An integer whose {@code hashCode()} implementation vastly reduces the risk of hash collisions in structured data
+ * using this type, when compared to the {@code java.lang.Integer} boxed type.
+ */
+public class EntityNum implements Comparable<EntityNum> {
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/services/utils/EntityNum.java:45]: Do we need this class?

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-web3/src/main/java/com/hedera/services/utils/EntityNum.java (hunk 5 above)
Comment (ivan-02) [reply hedera-mirror-web3/src/main/java/com/hedera/services/utils/EntityNum.java]: Removed

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/models/UniqueToken.java (hunk 4 above)
Comment (ivan-02) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/models/UniqueToken.java:32]: Done

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/models/UniqueToken.java (hunk 3 above)
Comment (ivan-02) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/models/UniqueToken.java]: Done

File: hedera-mirror-web3/build.gradle.kts [hunk 6]
```diff
@@ -47,6 +47,7 @@ dependencies {
     testImplementation(project(path = ":common", configuration = "testClasses"))
     testImplementation("com.playtika.testcontainers:embedded-postgresql")
     testImplementation("io.projectreactor:reactor-test")
+    testImplementation("org.mockito:mockito-inline:2.7.21")
```
[the remark this answers is not in the document]
Comment (ivan-02) [reply hedera-mirror-web3/build.gradle.kts]: Done

**Review by steven-sheehy — APPROVED** [review]
LGTM

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/models/UniqueToken.java (hunk 4 above)
Comment (steven-sheehy) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/models/UniqueToken.java:32]: I don't think we should be using Lombok in services copied code. We should try to keep this files as exact copies or as close as we can to its upstream services code so we can easily update it in the future.

Comment (ivan-02) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/models/UniqueToken.java:30]: In order to make the object immutable and keep the code as close as we can to services, I've removed the setters and made the fields final.

**Review by steven-sheehy — APPROVED** [review]

**Pull request merged**